### ISSUE TITLE
Lock inconsistencies cause deadlocks

### ISSUE THREAD

**Issue by borgbyte** [Countly/countly-sdk-cpp#85]
The function `Countly::sendHTTP` is sometimes called with `Countly::mutex` acquired.

This inconsistency can cause deadlocks as the custom HTTP client may need to lock some other mutex.

**Comment by ZahidZafar** [comment]
Hello @borgbyte,

You are right, there are inconsistencies in how the "Countly::sendHTTP" method handles mutexes. We will look into fixing that.
Currently, the SDK is designed around locking the mutex when making HTTP calls.

Could you share the specifics of the use case that was causing deadlocks for you?

In our testing, we could only produce a problem if other SDK calls (that try to acquire the mutex) were being called inside of the custom HTTP client method. Currently, this kind of use case is not supported.

We looked into producing a deadlock under other circumstances (while using a custom HTTP client), but we failed. Could you share any details or steps that would allow us to reproduce it?
Thanks,

**Comment by borgbyte** [comment]
> Currently, the SDK is designed around locking the mutex when making HTTP calls.

Deadlock aside, this is also a problem, because this means that calls to `Countly` instance can block a "bit" and cause the UI to stuck. The workaround is to defer these calls to a separate thread, but then what is the purpose of `start_thread=true`? ISTM these calls should be non blocking. I also don't understand why the mutex needs to be locked while the call is done.

On a separate note, would be nice to support asynchronous HTTP client - this could be done by passing a "done" callback that should be called after getting the HTTP response. For instance:

```cpp
using HTTPResponseCallback = std::function<void(HTTPResponseCallback)>;
using HTTPClientFunction = std::function<void(bool, const std::string&, const std::string&, HTTPResponseCallback)>;
```

Meanwhile, I'll try to detail how to cause the deadlock.

**Comment by borgbyte** [comment]
@ZahidZafar suppose the following:

- Tm - Main thread
- Tc - Countly thread

 - Mm - Main event loop mutex
 - Mc - Countly mutex

These are some cases of lock order while the app runs:
 - Tm -> lock Mm  (some HTTP request)
 - Tm -> lock Mm -> lock Mc  (record event)
 - Tc -> lock Mc -> [call http client] lock Mm -> (sending events to countly backend)

So you see that that Mc and Mm and locked in different orders in the last 2 cases, and this leads to deadlock. Ideally the last case would be
 - Tc -> [call http client] lock Mm -> (sending events to countly backend)

This roughly explains how a deadlock can happen, hope it helps.

**Comment by ZahidZafar** [comment]
In this issue you described 2 networking related issues that you were encountering:
 * calling SDK functions would block the main thread as we would try to make the network request on the same thread
 * there were chances of deadlocks due to the SDK not releasing its internal lock when sending requests to the server, using the custom HTTP client both of them should be fixed in the release we just published (version 22.06.0).In this version, we are adding requests into an internal queue before making an HTTP call.
SDK will process these requests one by one in a separate thread. To solve the deadlock issue, the SDK will now unlock its mutex while making network requests. You asked, "what is the purpose of 'start_thread=true'".
We use it as a flag to enable automatic session updates which would be happening on a separate thread.
